synapse/util/async_helpers.py

```diff
@@ -225,6 +225,18 @@ def __init__(self, name=None, max_count=1, clock=None):
             {}
         )  # type: Dict[str, Sequence[Union[int, Dict[defer.Deferred, int]]]]
 
+    def is_queued(self, key) -> bool:
+        """Checks whether there is a process queued up waiting
+        """
+        entry = self.key_to_defer.get(key)
+        if not entry:
+            # No entry so nothing is waiting.
+            return False
+
+        # There are waiting deferreds only in the OrderedDict of deferreds is
```

Comment:
```suggestion
        # There are waiting deferreds only if the OrderedDict of deferreds is
```

I don't think `entry` is an `OrderedDict`, isn't it a `Sequence`? I'm actually a little confused why this would check `entry[1]`.

Reply:
https://github.com/matrix-org/synapse/blob/a6a40a1519f3cef21935b7c69b444ae90fe1514a/synapse/util/async_helpers.py#L220-L226 explains it. The sequence is basically a data structure of "number of stuff current in flight" and "deferreds that are waiting".

Reply:
i.e. `entry[1]` is getting the `OrderedDict` of deferreds waiting.

Reply:
Ah, I see. I only read half the comment. 🙄 Should this really be a `Tuple[int, Dict[defer.Deferred, int]]` I wonder?

Reply:
We talked about this a bit...and it can't be a tuple cause it gets modified. Anyway, the type checking here is funky, but it is essentially a list of two items.